design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+- Did the toolkit get larger on iOS or Android?
+- Did the frame rate go down on either platform?
+- Is the toolkit taking longer to startup on either platform?
+
+If the answer to any of these questions is 'yes', then we move to the next stage.
+
+#### Required Changes
+
+These are contingent on what we may have made worse and why.  Opportunities to improve that specific situation will be strategic and hard to pin down at this point.  A lot depends on why the toolkit is bigger, slower, etc...  So this is speculation right now.
+
+It depends on the other changes, but let's break it down by type.
+- Is the toolkit slower to render?  Look at the frame logic and do more up front or on other threads.
+- Is the toolkit slower to startup?  Figure out why and consider doing more work on separate threads or later.
+- Is the toolkit bigger?  Look at slicing unused layers out with compile flags, getting rid of unused code or dependencies.
+
+#### Benefits
+
+The main benefit is just keeping the performance the same.  When we implement Metal on iOS we can expect rendering performance to get much better, of course, but that's not the goal of this effort.  We aim to not make it any worse while reaching our other goals.
+
+While we're doing that, we may run across areas for improvement.  I'd suggest that the developers write these up as they find them.  Some obvious improvements are:
+- Slicing out unused functionality
+- Making more specific functionality optional
+- Reducing or eliminating dependencies on third party libraries.
+
+There's a lot that could be done in general and I'd encourage sponsors to pay maintainers to look into those.
+
+## API Modifications
+
+At present we do not expect to change any existing APIs at the public level.  We do expect to add some new API methods and classes.  We would detail these as we proceeded with the implementation.
+
+## Migration Plan and Compatibility
+
+We do not intend for these changes to be incompatible with the existing version.
+
+## Rejected Alternatives
+
+The main alternative would be switching to an intermediate toolkit to implement both OpenGL and Metal.  The problem with that approach is toolkit size.  Quite a few users cannot abide with the toolkit growing considerably.
+
+## Pull Request Schedule
+
+At the request of the community we've put together a list of Pull Requests that we would organize the work into.  These would be individual pieces that would leave the toolkit in working shape and set us up for the next PR.
+
+#### PR 1: Shaders
+
+This PR will address shader flexibility.
+
+Sections covered:
+* Shader Representation Changes
+* Shader Registry
+
+By the end we should have:
+* Shaders called out as individual objects accessible outside the toolkit
+* Shaders create-able from outside the toolkit
+* Shaders replaceable from outside the toolkit
+* A compile time option to use shaders either from the compressed source (as now) or with individual shader source files
```

Comment:
The way it is now, compressed shader source/header code generation is done using a javascript utility. Do you have plans to create a new generator for this?

Reply:
My plan was to leave that route possible.  I suspect some people have workflows that depend on it.
But then to also rip the shaders out as source and make the new route possible.  I guess we might want to make the tooling understand that, writing the shaders somewhere we can pick them up.

With the Metal version (next PR, of course) we're not going to try to generate those from the GLSL.  We'll have Metal versions and have to explicitly deal with divergence.